### PR TITLE
fix `unnecessary_parenthesis` false positive for function expressions

### PR DESCRIPTION
**Pull request by pq**
Fixes #https://github.com/dart-lang/linter/issues/2944

/cc @bwilkerson @srawlins

### CODE REVIEW

File: lib/src/rules/unnecessary_parenthesis.dart [hunk 1]
```diff
@@ -68,6 +68,12 @@ class _Visitor extends SimpleAstVisitor<void> {
       return;
     }
 
+    // https://github.com/dart-lang/linter/issues/2944
+    // (() => foo()).bar();
+    if (expression is FunctionExpression && parent is MethodInvocation) {
```
Comment (srawlins) [inline lib/src/rules/unnecessary_parenthesis.dart:73]: In addition to MethodInvocation this same problem may happen for (deep breath):

* getter access: `(() => foo()).bar;`
* setter access: `(() => foo()).bar = 7;`
* operators: `(() => foo())[6];` or `(() => foo())++;`
* type instantiation: `(() => foo()).bar<int>;`

Comment (pq) [reply lib/src/rules/unnecessary_parenthesis.dart:73]: Right!  Good call.  I'll update.

**Comment by pq** [comment]
@srawlins: I think I got the cases covered.  PTAL?

File: test_data/rules/unnecessary_parenthesis.dart [hunk 2]
```diff
@@ -11,13 +11,21 @@ import 'dart:async';
...
+  (() => '')[0]; // OK
 }
 
 extension A on void Function() {
   void Function() compose(void Function() other) => () {
     this();
     other();
   };
+
+  set f(int f) {}
```
Comment (srawlins) [inline test_data/rules/unnecessary_parenthesis.dart:26]: This stuff is a trip. extension on a function type... 🤯

**Review by srawlins — APPROVED** [review]

**Review by bwilkerson — APPROVED** [review]

**Pull request merged**